src/Java.Interop.Tools.JavaTypeSystem/Adapters/ManagedApiImporter.cs

```diff
@@ -0,0 +1,379 @@
+				javaParentType: parent,
+				deprecated: deprecated,
+				jniSignature: jni_signature.ToString (),
+				isSynthetic: false,
+				isBridge: false,
+				returnJni: jni_signature.Return.Jni,
+				isNative: false,
+				isSynchronized: false,
+				returnNotNull: false
+			);
+
+			for (var i = 0; i < jni_signature.Parameters.Count; i++)
+				model.Parameters.Add (ParseParameterModel (model, jni_signature.Parameters [i], method.Parameters [i]));
+
+			return model;
+		}
+
+		static JavaParameterModel ParseParameterModel (JavaMethodModel parent, JniTypeName jniParameter, ParameterDefinition managedParameter)
+		{
+			var raw_type = jniParameter.Type;
+
+			// This covers a special case where we have generated an interface method like:
+			// void DoThing (System.Collections.Generics.IList<MyPackage.MyType> p0);
+			// However the [Register] JNI signature for this method is missing generic information:
+			// "(Ljava/util/List;)V"
+			// We need to try to rebuild the generic signature from the [Register] attributes
+			// on the type components that make up the signature:
+			// java.util.List<mypackage.MyType>
+
+			// TODO: This is more correct, but differs from ApiXmlAdjuster.
+			//if (managedParameter.ParameterType is GenericInstanceType)
+			//	if (TypeReferenceToJavaType (managedParameter.ParameterType) is string s)
+			//		raw_type = s;
+
+			return new JavaParameterModel (parent, managedParameter.Name, raw_type, jniParameter.Jni, false);
+		}
+
+		static void AddReferenceTypeRecursive (JavaTypeModel type, JavaTypeCollection collection)
+		{
+			collection.AddReferencedType (type);
+
+			foreach (var nested in type.NestedTypes)
+				AddReferenceTypeRecursive (nested, collection);
+		}
+
+		static bool ShouldSkipType (TypeDefinition type)
+		{
+			// We want to use Java's collection types instead of our managed adapter.
+			// eg: 'Java.Util.ArrayList' over 'Android.Runtime.JavaList'
+			// So don't import our adapters.
+			switch (type.FullName) {
+				case "Android.Runtime.JavaCollection":
+				case "Android.Runtime.JavaDictionary":
+				case "Android.Runtime.JavaList":
+				case "Android.Runtime.JavaSet":
+					return true;
+			}
+
+			// Currently we do not support generic types because they conflict.
+			// ex: AdapterView`1 and AdapterView both have:
+			// [Register ("android/widget/AdapterView")]
+			// So we do not import the generic type if we also find a non-generic type.
+			var non_generic_type = type.HasGenericParameters
+				? type.Module.GetType (type.FullName.Substring (0, type.FullName.IndexOf ('`')))
+				: null;
+
+			if (ShouldSkipGeneric (type, non_generic_type, null))
+				return true;
+
+			return false;
+		}
+
+		static bool ShouldSkipGeneric (TypeDefinition? a, TypeDefinition? b, TypeDefinitionCache? cache)
+		{
+			if (a == null || b == null)
+				return false;
+			if (!a.ImplementsInterface ("Android.Runtime.IJavaObject", cache) || !b.ImplementsInterface ("Android.Runtime.IJavaObject", cache))
```

Comment:
I think we can/should start using `IJavaPeerable` instead of `IJavaObject`, as that'll better support eventual "desktop JVM" scenarios.

Reply:
Let's do that separately.  This PR is big enough and hard enough to ensure nothing is breaking already.